decisions/0004-remixing-react-router.md

```diff
@@ -0,0 +1,201 @@
+  - Instead when an action redirects, we'll use a PUSH and in this case the history stack would become `[/, /login, /private]` and the user would be able to navigate back through the login page and to the home page
+
+Note: User's can still be explicit here and use `<Form method="post" replace={shouldReplace}>` and the router will respect the value passed to `replace`.
+
+### `unstable_shouldReload` stabilized as `shouldRevalidate`
+
+We stabilized the API for when a given route loader should re-run, and changed the name to align with the "revalidation" nomenclature and the `useRevalidator` hook. We also leave more control in the hands of the user here. In Remix there were some cases in which you _could not_ opt out of revalidation and if your method did run you had full control and couldn't necessarily handle one edge case and then say "do what you otherwise would have done".
+
+Now, if you provide a `shouldRevalidate` method we will call it during all revalidations and provide you a `defaultShouldRevalidate` boolean value. This allows you to opt out of any revalidation, and also code your own logic to fallback on our default choice:
+
+```js
+function shouldRevalidate({ defaultShouldRevalidate }) {
+  // Don't revalidate for this case
+  if (someEdgeCase()) {
+    return false;
+  }
+
+  // Otherwise, do what we would have done by default
+  return defaultShouldRevalidate;
+}
+```
+
+### `<ScrollRestoration getKey>` prop
+
+In Remix, the `<ScrollRestoration>` component made an assumption that we would always restore scroll position based on `location.key`. If the key was the same as a prior location we knew the scroll position for, then we knew you had been there before and we should restore. This works great for back/forward navigations but it's a bit overly restrictive. You cannot choose to restore scroll based on anything other than `key`. Twitter has a great implementation of this as you click around in their left nav bar - your tweet feed is always at the same place when you click back to it - even though it's a _new_ location in the history stack. This is because they're restoring by pathname here instead of `location.key`. Or maybe you want to maintain scroll position for all routes under a given pathname and you thus want to use a portion of the pathname as the scroll restoration key.
+
+In React Router we now accept an optional `<ScrollRestoration getKey>` prop where you provide a function that returns the key to use for scroll restoration:
+
+```js
+function getKey(location: Location, matches: DataRouteMatch[]) {
+  // Restore by pathname on /tweets
+  if (location.pathname === "/tweets") {
+    return location.pathname;
+  }
+  // Otherwise use the key
+  return location.key;
+}
+```
+
+**Backwards Compatibility**
+
+We're ok here since the new prop is optional and defaults to using `location.key`
+
+### `<Link resetScroll="false">` prop
```

Comment:
Why not `<Link preventScrollReset>` or something like that? With resetScroll the default behavior will be true so you will have to do `<Link resetScroll={false} />` if the prop default is false the usage would be just `<Link propName />` which is a little nicer.

Reply:
@ryanflorence What do you think?